subsys/caf/modules/ble_adv.c

```diff
@@ -306,153 +315,303 @@ static int ble_adv_start_undirected(bool fast_adv)
+	case STATE_DISABLED_OFF:
+		return "DISABLED_OFF";
 
-		k_work_cancel_delayable(&adv_update);
-		k_work_reschedule(&grace_period_end, K_SECONDS(used_grace_period));
+	case STATE_OFF:
+		return "OFF";
 
-		state = STATE_GRACE_PERIOD;
-	} else if (err == -EAGAIN) {
-		LOG_INF("No active advertising");
-		err = ble_adv_stop();
-		if (!err) {
-			state = STATE_OFF;
-			module_set_state(MODULE_STATE_OFF);
-		}
+	case STATE_IDLE:
+		return "IDLE";
+
+	case STATE_ACTIVE:
+		return "ACTIVE";
+
+	case STATE_DELAYED_ACTIVE:
+		return "DELAYED_ACTIVE";
+
+	case STATE_GRACE_PERIOD:
+		return "GRACE_PERIOD";
+
+	case STATE_ERROR:
+		return "ERROR";
+
+	default:
+		/* Should not happen. */
+		__ASSERT_NO_MSG(false);
+		return NULL;
+	}
+}
+
+static void update_state_internal(enum state new_state)
+{
+	if (new_state == STATE_ACTIVE) {
+		fast_adv = (req_fast_adv && IS_ENABLED(CONFIG_CAF_BLE_ADV_FAST_ADV));
+		direct_adv = can_direct_adv(cur_identity);
+	} else {
+		fast_adv = false;
+		direct_adv = false;
+	}
+
+	grace_period_s = req_grace_period_s;
+
+	state = new_state;
+
+	LOG_DBG("State: %s fast_adv: %d direct_adv: %d grace_period: %zu[s]",
+		state2str(state), fast_adv, direct_adv, grace_period_s);
+}
+
+static void update_grace_period_work(void)
+{
+	if (!IS_ENABLED(CONFIG_CAF_BLE_ADV_GRACE_PERIOD)) {
+		return;
+	}
+
+	if (state == STATE_GRACE_PERIOD) {
+		__ASSERT_NO_MSG(grace_period_s > 0);
+		(void)k_work_reschedule(&grace_period_end, K_SECONDS(grace_period_s));
 	} else {
-		LOG_ERR("Cannot modify advertising data (err %d)", err);
+		(void)k_work_cancel_delayable(&grace_period_end);
 	}
+}
 
-	return err;
+static void update_fast_adv_work(void)
+{
+	if ((state == STATE_ACTIVE) && fast_adv && !direct_adv) {
+		(void)k_work_reschedule(&fast_adv_end,
+					K_SECONDS(CONFIG_CAF_BLE_ADV_FAST_ADV_TIMEOUT));
+	} else {
+		(void)k_work_cancel_delayable(&fast_adv_end);
+	}
 }
 
-static void ble_adv_update_fn(struct k_work *work)
+static void notify_ble_stack(void)
 {
-	bool can_fast_adv = false;
+	(void)k_work_cancel(&adv_delayed_start);
 
 	switch (state) {
-	case STATE_DELAYED_ACTIVE_FAST:
-		can_fast_adv = true;
+	case STATE_DISABLED:
+	case STATE_DISABLED_OFF:
+		break;
+
+	case STATE_OFF:
+	case STATE_IDLE:
+	case STATE_ERROR:
+		ble_adv_stop();
+		break;
+
+	case STATE_DELAYED_ACTIVE:
+		ble_adv_start_schedule();
 		break;
 
-	case STATE_ACTIVE_FAST:
-	case STATE_DELAYED_ACTIVE_SLOW:
+	case STATE_ACTIVE:
+	case STATE_GRACE_PERIOD:
+		ble_adv_start();
 		break;
 
 	default:
 		/* Should not happen. */
 		__ASSERT_NO_MSG(false);
 	}
+}
+
+static void update_state(enum state new_state)
+{
+	if (state == STATE_ERROR) {
+		return;
+	}
 
-	int err = ble_adv_start(can_fast_adv);
+	enum state prev_state = state;
 
-	if (err) {
-		module_set_state(MODULE_STATE_ERROR);
+	update_state_internal(new_state);
+	broadcast_module_state(prev_state, state);
+
+	notify_ble_stack();
+	update_grace_period_work();
+	update_fast_adv_work();
```

Comment:
almost everything in this file has it the other way round (leading verb) - do you really want to change it?
I have not checked but I expect that there is no consistence in NCS regarding this.

Reply:
In other files, we also often used verb at the beginning (e.g. event handlers are called `handle_..._event`

Reply:
Well, @pdunaj, you are correct there is no consistency regarding this in NCS, but most API functions have a verb at the end. We may want to agree on the same naming approach so that NCS has the same touch and feel to the user and does not like a bundle of independent modules that do not share any common philosophy.

Anyway, that's a topic for a separate discussion and should not be part of this PR.